[3.x] Physics Interpolation - Add `InterpolatedProperty`

lawnjelly opened the pull request:
Adds framework for `InterpolatedProperties`.
Adds interpolated properties to `Camera`.

Interpolates:
* Perspective (fov, near, far)
* Orthogonal (size, near, far)
* Frustum (size, frustum_offset, near, far)

Implements https://github.com/godotengine/godot-proposals/issues/9964 .

## Notes
* The move from server to scene side FTI makes this far easier.
* `Camera` provides a good test case for adding some custom interpolated properties.
* We could do all of it manually, but it makes some sense to create a helper template to reduce boiler plate.
* The `InterpolatedProperty` emulates the underlying type (for when FTI is off), but has some extra functionality available.
* Standardizes the use of `fti_pump()` for more reset functionality in `Spatial` derived classes.
* Automatically responds to `reset_physics_interpolation()` user calls.

## Discussion
While implementing this I made some minor adjustments to `SceneTreeFTI` to better support properties.

Properties aren't inherited and concatenated like scene tree xforms, so it turned out these are more efficient to handle separately.

* They only need to be updated once per frame (at the end) instead of two passes.
* They can maintain their own frame update list, instead of checking each node in the `SceneTree`, as often only a few nodes will need properties interpolating.
* Nodes when notifying changes now indicate via a bool whether the changes are to the transform, or to properties. In most cases transforms are already handled by the `Spatial` base class, so in most custom cases in the future contributors will be notifying custom property changes.

### `fti_pump` and `fti_update_servers`
One slightly controversial change it that `fti_pump()` and `fti_update_servers` is now separated into two versions - for xforms and for properties.

As far as I can see this is the most efficient way of handling it, even if it may look slightly more confusing in say the `Camera` class, where we have to update servers for _both_ xform and properties.

An alternative would be to send e.g. a `bool` with the two routines, however there is more opportunity to "shoot yourself in the foot" with that approach, particularly regarding calling the parent class routines.

In practice, most classes that custom interpolate will only be updating properties, and will rely on `Spatial` for xforms, so will only have to implement `fti_pump_property()` and `fti_update_servers_property()` - `Camera` is to some extent an anomaly and a worst case situation, so good to handle first.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

lawnjelly committed:
Physics Interpolation - Add InterpolatedProperty

And add some basic interpolated properties to Camera.

lawnjelly commented on core/error_macros.h:
```diff
@@ -552,10 +552,14 @@ void _physics_interpolation_warning(const char *p_function, const char *p_file,
+			first_print = false;                                                                                         \
+		}                                                                                                                \
+	} else                                                                                                               \
 		((void)0)
```
Note: I just changed this to be slightly more efficient in the case of slow checks.

It now checks the static _before_ doing the condition, so once it has triggered the first time, it no longer does the condition check.

rburing commented on core/math/transform_interpolator.h:
```diff
@@ -84,6 +86,15 @@ class TransformInterpolator {
 	static real_t checksum_transform(const Transform &p_transform);
 	static Method find_method(const Basis &p_a, const Basis &p_b);
+
+	// Wrappers for InterpolatedProperty, so it can use various standard interpolation methods and reduce boilerplate.
+	static float interpolated_property_lerp(float p_a, float p_b, float p_fraction) {
+		return Math::lerp(p_a, p_b, p_fraction);
+	}
+
+	static Vector2 interpolated_property_lerp(const Vector2 &p_a, const Vector2 &p_b, float p_fraction) {
+		return p_a.linear_interpolate(p_b, p_fraction);
+	}
```
These are a bit out of place here, no? Why not put them inside `InterpolatedProperty`?

lawnjelly replied:
That would also be fine (and I can see that makes a lot of sense, and I'm not averse to sticking them there at least temporarily, they are pretty easy to relocate at this stage, so I can stick them there for now).

I remember being unsure the most appropriate place to put these longterm, and it wasn't cut and dry. Probably between:

* `InterpolatedProperty`
* `TransformInterpolator`
* `Math`

I admit there was some pre-empting here - I can foresee future demand for some kind of user mechanism to do the same from gdscript, or using these from other areas, in which case if these are in `InterpolatedProperty` we'll probably want to move them somewhere else more generic.

I did consider the alternative of putting them directly in `Math`, but was aware the advice is usually to localize solutions first, and _then_ consider making them more generic only when required. See [best practices for engine contributors](https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html#prefer-local-solutions) .

`TransformInterpolator` is currently just a way of conveniently grouping all the physics interpolation functions together at the moment (without a lot of dependencies), so they can be used from different call points (client, server, bound funcs). Some of the existing functionality is likely a little too specific to put in `Math`. The other divergence is that `Math` is intended to be generic, whereas `TransformInterpolator` is code to do a specific task fast (and there may be shortcuts in terms of error checking where done in advance).

Regarding lerp functions specifically, there is a general problem to solve here that they are mostly OOP functions in a bunch of the classes, like `Vector2` `Vector3` etc, but they are not standardized (especially in 3.x), and calling them from macros that don't care about the underlying class is not always easy / possible. This may well have come up before in 4.x too about gdscript, in the discussions about the bound free function `lerp` versus OOP functions.

I suspect in the longrun we may end up moving them (or similar) to `Math`, but more bikeshedding will be involved there and sometimes it can be nice to defer this until a generic need is required, as the guidelines say.

rburing replied:
I think the way to follow the guideline would be to put them in `InterpolatedProperty`, not some other class which is currently dedicated specifically to transforms. I understand that in the future this lerp code and the TransformInterpolator code could be put together somewhere with bindings, but imho it makes more sense to move things around (and rename the class) then instead of now. There could be a TODO comment about this unification (and possibly binding). Anyway, I'll leave it up to you.

rburing reviewed (approved):
This looks quite nice from the API perspective, and I appreciate the defensive dev checks.

I'll have to defer to you for the implementation. I can say it looks believable.

Ivorforce commented on core/interpolated_property.h:
```diff
@@ -0,0 +1,102 @@
+	T prev;
+
+public:
+	void pump() {
```
I'm not familiar with the term "pump" in this context. Is it a common name for this operation?
How about adding a comment? Or maybe a more descriptive name like `snap_to_current_value` / `jump_to_current_value`?

lawnjelly replied:
Naming is always one of the most difficult problems in programming :grin: .. I've tried a few different names for this over the years but currently am tending towards `pump` or `fti_pump`. It essentially acts like a pump, on every tick you are moving current values over to previous values (it is also sometimes reused for reset, but resetting refers to the _intention_ in a specific circumstance, not the _mechanism_).

_(This is also why `teleport` is imo a better naming than `reset_physics_interpolation`, because it makes the intention clear, whereas `reset` is very vague and users often get confused about it.)_

This pump is the core of how fixed timestep interpolation works, it relies on maintaining a continuous flow of previous and current transforms / properties / whatever, and then interpolating between them using the interpolation fraction.

I'm happy to put some more comments in.

`snap_to_current_value` / `jump_to_current_value` I'm not sure describe what is going on, I for one have no idea what they are referring to (and I've been using FTI for 25 years lol :grin: ).

lawnjelly replied:
Actually thinking about it, it may be possible to use `fti_tick()` here, let me have a look through.

Ivorforce commented on core/interpolated_property.h:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	bool operator==(const T &p_o) const {
+		return p_o == *this;
```
I'm not a big fan of the implicit conversion here. Might be easier to understand with explicit conversions:

```suggestion
		return InterpolatedProperty(p_o) == *this;
```

lawnjelly replied:
Ah good point. In fact I think it may be the other way around, we want to compare them as `T`, let me check. :+1:

Ivorforce commented on core/interpolated_property.h:
```diff
@@ -0,0 +1,102 @@
+	void set_interpolated_value(const T &p_val) {
+		_interp = p_val;
+	}
+	const T &interp() const { return _interp; }
```
```suggestion
	const T &interpolated_value() const { return _interp; }
```

I'm not a big fan of abbreviated names for functions :)

lawnjelly commented:
Have pushed some changes in response to the suggestions:
* Moved the `lerp` functions to `InterpolatedProperty` (actually in a namespace so they could be in the cpp to avoid header includes, as `InterpolatedProperty` itself is a template)
* Fixed up the `==` and `!=` operators
* Added some more comments to the `pump` functions
* Changed `interp()` to `interpolated()`

lawnjelly commented:
Thanks!
